### PR TITLE
Correct minimum RPM version support

### LINKED ISSUE #2289: "pbench-sysstat-12.0.3 is not installed" showing in metadata.log after running pbench-user-benchmark
With the latest build of pbench-agent https://copr.fedorainfracloud.org/coprs/portante/pbench-test/build/2162023/, after running pbench-user-benchmark, metadata.log shows `pbench-sysstat-12.0.3 is not installed`, even though it is installed. The logs are below.
```
pbench-user-benchmark -- sleep 60
pbench-tool-data-sink: connected to redis server localhost:17001
pbench-tool-meister: connected to redis server localhost:17001
Running sleep 60 for iteration 1-default
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
pidstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/pidstat', missing
iostat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/iostat', missing
mpstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/mpstat', missing
perf: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/perf', missing
pidstat: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/pidstat', missing
proc-interrupts: post-processing data
proc-vmstat: post-processing data
sar: WARNING - postprocess - tool output directory, '/var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/1-default/sample1/tools-default/ip-172-31-94-3.ec2.internal/sar', missing
Collecting system information
```
```
dnf install -y pbench-sysstat
Updating Subscription Management repositories.
Last metadata expiration check: 0:01:04 ago on Thu 06 May 2021 12:17:45 PM UTC.
Package pbench-sysstat-12.0.3-1.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```
```
cat /var/lib/pbench-agent/pbench-user-benchmark__2021.05.06T12.19.04/metadata.log
[pbench]
config = 
date = 2021-05-06T12:19:04
name = pbench-user-benchmark__2021.05.06T12.19.04
rpm-version = v0.71.0-171gfe0675763
script = pbench-user-benchmark
iterations = 1-default

[controller]
hostname = ip-172-31-94-3.ec2.internal
hostname-s = ip-172-31-94-3
hostname-f = ip-172-31-94-3.ec2.internal
hostname-i = 172.31.94.3
hostname-a = ip-172-31-94-3.ec2.internal
ssh_opts = -o StrictHostKeyChecking=no

[run]
controller = ip-172-31-94-3.ec2.internal
start_run = 2021-05-06T12:19:05.489913
end_run = 2021-05-06T12:20:25.934423

[tools]
hosts = ip-172-31-94-3.ec2.internal
group = default
trigger = None

[tools/ip-172-31-94-3.ec2.internal]
label = 
tools = iostat,mpstat,perf,pidstat,proc-interrupts,proc-vmstat,sar,turbostat
hostname-s = ip-172-31-94-3
hostname-f = ip-172-31-94-3.ec2.internal
hostname-i = 172.31.94.3
hostname-a = ip-172-31-94-3.ec2.internal
rpm-version = v0.71.0-171gfe0675763
iostat = --interval=3
mpstat = --interval=3
perf = --record-opts='-a --freq=100'
pidstat = --interval=30
proc-interrupts = --interval=3
proc-vmstat = --interval=3
sar = --interval=3
turbostat = --interval=3

[tools/ip-172-31-94-3.ec2.internal/iostat]
options = --interval=3
install_check_status_code = 1
install_check_output = iostat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/mpstat]
options = --interval=3
install_check_status_code = 1
install_check_output = mpstat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/perf]
options = --record-opts='-a --freq=100'
install_check_status_code = 1
install_check_output = perf: perf is not installed

[tools/ip-172-31-94-3.ec2.internal/pidstat]
options = --interval=30
install_check_status_code = 1
install_check_output = pidstat: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/proc-interrupts]
options = --interval=3
install_check_status_code = 0
install_check_output = 

[tools/ip-172-31-94-3.ec2.internal/proc-vmstat]
options = --interval=3
install_check_status_code = 0
install_check_output = 

[tools/ip-172-31-94-3.ec2.internal/sar]
options = --interval=3
install_check_status_code = 1
install_check_output = sar: pbench-sysstat-12.0.3 is not installed

[tools/ip-172-31-94-3.ec2.internal/turbostat]
options = --interval=3
install_check_status_code = 0
install_check_output = turbostat: kernel-tools is installed

[iterations/1-default]
iteration_number = 1
iteration_name = 1-default
user_script = sleep, 60
```

### PR DESCRIPTION
**Pull request by portante**
This fixes #2289, a bug that was introduced in PR #2234 (commit 5794fc1).

This is a combination of two main problems.  The first was that we incorrectly mocked the output of `rpm --query ...`.  It only returns EVRs, which are version strings, which explicitly don't include the queried RPM name.  Additionally, we were not formatting the output with a new line, so for RPMs that have multiple versions installed, the code would not work correctly.

The second problem was that our `pbench-sysstat` requirement was for `12.0.3`, but the RPMs we have built publicly have version `12.0.3-1`, which won't match exactly.

We have also corrected the `rpm` mock for the bench-scripts to behave more closely to, but not exactly as, the actual `rpm` behavior.

----

The second commit refactors the use of `check_install_rpm` from the `base` script.

The `check_install_rpm` was only being used by the `bench-scripts`, and at that, they all almost did the same thing as a result of its execution.  So instead of having layers on layers, we have combined all the layers into one, removing much duplication.  This also has each script record its name in the logs instead of having the name of the common script itself.

This also removes duplicate reports regarding RPMs in the logs.

### CODE REVIEW

File: agent/tool-scripts/base-tool
Comment (webbnh) [inline agent/tool-scripts/base-tool:110]: As a "quick fix", this is fine; however, in general, it's annoying:  this is a weird place to enshrine this information (e.g., why isn't it in some sort of `requirements.txt` kind of place?), and you've got it in _two_ places -- here and at line 67 -- and I'm pretty sure that we don't want them to get out of sync (so, they should probably have a common definition which is referenced here).

File: agent/tool-scripts/base-tool
Comment (webbnh) [inline agent/tool-scripts/base-tool:579]: Um...why is this function definition inside a `case` branch?  (Thanks for renaming it, btw.  ;-) )

**Review by ndokos — DISMISSED** [review]
LGTM

**Review by webbnh — APPROVED** [review]
As a quick fix, this is fine, but burying the `sysstat` version requirement partway down inside a script is ungood.

You should consider capturing the `sysstat` version requirement in a local variable defined at the top of the file and referenced where needed.

File: agent/tool-scripts/base-tool
Comment (webbnh) [reply agent/tool-scripts/base-tool:579]: OK, nevermind:  it's defined there because all of the references to it are in that same branch (the _nested_ `case` confused me...still, it's a bit weird to have a function definition inside a nested structure, no?).

File: agent/tool-scripts/base-tool
Comment (portante) [reply agent/tool-scripts/base-tool:110]: Agreed on all points.  This is something that we can address as we convert from `bash` to `Python3`, and by dropping the `sysstat` tools in favor of PCP.

File: agent/tool-scripts/base-tool
Comment (portante) [reply agent/tool-scripts/base-tool:579]: Because it is only needed for `--install`?

File: agent/bench-scripts/test-bin/rpm
Comment (webbnh) [inline agent/bench-scripts/test-bin/rpm:24]: This doesn't look right:  in two of the cases you're comparing `${3}` to a tool name/package and returning a version number; in the third case (line 23), I presume that you're going to return the tool/package rather than just a version number...not so?

Comment (webbnh) [reply agent/bench-scripts/test-bin/rpm:24]: Should it instead by erroring-out, calling for a developer to add another special case?

**Review by webbnh — APPROVED** [review]
If you're good with the below, then I'm good with it....

**Review by ndokos — APPROVED** [review]

File: agent/bench-scripts/test-bin/rpm
Comment (ndokos) [reply agent/bench-scripts/test-bin/rpm:24]: Yes, this does look a bit off: the real rpm would produce an error message for an unknown package and return status 1. I guess it depends on how closely the mocked version should resemble the original in its behavior. The mocked one basically pretends that the package always exists (by always returning status 0) and prints out some conventional "version" (which in some cases, e.g. dpdk-tools, is just the name of the package): the only thing that matters in that case is that the gold file has the same thing in it.

Comment (webbnh) [reply agent/bench-scripts/test-bin/rpm:24]: > the only thing that matters in that case is that the gold file has the same thing in it.

:-D

> the real rpm would produce an error message for an unknown package and return status 1.

The problem is, this mock has no idea of whether the unrecognized package exists or not, so it doesn't know whether to return an error or a mock'd version number.

I'm OK with the mock always pretending that an unrecognized package exists and returning success, but I think it needs to print out something intelligent (or, at least, intelligible), and I'm concerned that this isn't "it".

Comment (portante) [reply agent/bench-scripts/test-bin/rpm]: Resolved by making the mock'd `rpm` behavior as close to the real `rpm` as possible.

**Review by webbnh — DISMISSED** [review]
Once again, I won't hold up the merge for this if you disagree, but I think `test-bin/rpm` needs some additional tweaking.

File: agent/bench-scripts/test-bin/rpm
Comment (portante) [reply agent/bench-scripts/test-bin/rpm]: Fixed.

**Review by portante — COMMENTED** [review]
Ready for another pass.

The 2nd commit is intended to be kept separate.

All the comments on the first commit have been applied.

The reason for the second commit is that when debugging why the Tufts Univ COMP 150 DCC demo was not working in their VM environment, we had messages which were generic being emitted, and we didn't know where they were coming from.  Once we debugged and found them, it became clear that the `base` script was emitting messages without attributing them to the script it was working under.

Once we corrected that, I noticed a number of now (almost) duplicate messages, and refactored the code to simplify the logic.

The `comp-150` branch is depending on this work now.

**Review by webbnh — DISMISSED** [review]
Looks OK to me, modulo the one small question.

File: agent/base
Comment (portante) [reply agent/base]: Intentional, as this is really an internal error of sorts, not an error intended for the user.

**Review by portante — COMMENTED** [review]
Okay, all set now; that is to say, I one last change to be clear the error message is an internal error of sorts.

**Review by webbnh — DISMISSED** [review]

**Review by ndokos — APPROVED** [review]

**Review by riya-17 — APPROVED** [review]
Looks Good

**Review by webbnh — APPROVED** [review]

**Review by riya-17 — APPROVED** [review]

**Review by ndokos — APPROVED** [review]
I might not have clicked the button...

**Review by webbnh — APPROVED** [review]
Still good.

**Pull request merged**
